### PR TITLE
BUG:Pivot table drops column/index names=nan when dropna=false

### LINKED ISSUE #14072: pivot_table margins bottom-left total does not correspond to other content when dropna=False
#### Code Sample, a copy-pastable example if possible

In:
M = pd.DataFrame([
                  [1, 'a', 'A'], 
                  [1, 'b', 'B'], 
                  [1, 'c', None]], 
                 columns=['x', 'y', 'z'])
P = M.pivot_table(values='x', index='y', columns='z', aggfunc='sum', fill_value=0, margins=True, dropna=False)
P

Out:

z	|A	|B	|All
---|---|---|---
y|||			
a	|1.0	|0.0	|1.0
b	|0.0	|1.0	|1.0
All	|1.0	|1.0	|3.0

#### Expected Output
Either

z	|A	|B	|All
---|---|---|---
y|||			
a	|1.0	|0.0	|1.0
b	|0.0	|1.0	|1.0
All	|1.0	|1.0	|2.0

or

z	|A	|B	|None	|All
---|---|---|---|---
y||||			
a	|1.0	|0.0	|0.0	|1.0
b	|0.0	|1.0	|0.0	|1.0
c	|0.0	|0.0	|1.0	|1.0
All	|1.0	|1.0	|1.0	|3.0

depending on dropna's meaning for this case

#### output of ``pd.show_versions()``
INSTALLED VERSIONS
------------------
commit: None
python: 3.4.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.2.0-36-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.1
nose: 1.3.1
pip: 8.1.2
setuptools: 22.0.5
Cython: 0.20.1post0
numpy: 1.11.0
scipy: 0.17.1
statsmodels: None
xarray: None
IPython: 4.2.0
sphinx: None
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: None
tables: None
numexpr: 2.6.0
matplotlib: 1.5.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.5.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None

### PR DESCRIPTION
**Pull request by OXPHOS**
- [x] closes #14072
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [ ] whatsnew entry
clean-up of #14246

### CODE REVIEW

File: pandas/core/categorical.py [hunk 1]
```diff
@@ -548,10 +548,6 @@ def _validate_categories(cls, categories, fastpath=False):
 
         if not fastpath:
 
-            # Categories cannot contain NaN.
```
Comment (TomAugspurger) [inline pandas/core/categorical.py]: Do you have some unintentional changes in here? This shouldn't be removed.

Comment (OXPHOS) [inline pandas/core/categorical.py]: I feel like the Index, `Index([None, u'A', u'B'], dtype='object')` needs to be passed to `Categorical` when doing `MultiIndex`, as when `dropna=False`, `None` could also be the index/column name. Or I didn't get this correctly?

File: pandas/core/reshape/pivot.py [hunk 2]
```diff
@@ -159,15 +159,15 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
     if isinstance(table, DataFrame):
         table = table.sort_index(axis=1)
 
-    if fill_value is not None:
-        table = table.fillna(value=fill_value, downcast='infer')
-
     if margins:
         if dropna:
```
Comment (TomAugspurger) [inline pandas/core/reshape/pivot.py]: If I remove this `if dropna`, most of the tests pass (including a fix for this one) other than

```python
        df = pd.DataFrame({'a': [1, 2, 2, 2, 2, np.nan],
                           'b': [3, 3, 4, 4, 4, 4]})
        actual = pd.crosstab(df.a, df.b, margins=True, dropna=False)
        expected = pd.DataFrame([[1, 0, 1], [1, 3, 4], [2, 4, 6]])
        expected.index = Index([1.0, 2.0, 'All'], name='a')
        expected.columns = Index([3, 4, 'All'], name='b')
```

Here's the result and expected

```
(Pdb) pp actual
b    3  4  All
a
1.0  1  0    1
2.0  1  3    4
All  2  3    5
(Pdb) pp expected
b    3  4  All
a
1.0  1  0    1
2.0  1  3    4
All  2  4    6
```

You have more experience with this section of the code than I do, but the margins on the `expected` look incorrect to me.

**Comment by jreback** [comment]
this is doing similar things to changed in #12607

**Comment by TomAugspurger** [comment]
> this is doing similar things to changes in #12607

Ah I see. That is a much larger change that the original issue I was looking at :)

File: pandas/core/reshape/pivot.py (hunk 2 above)
Comment (OXPHOS) [inline pandas/core/reshape/pivot.py]: You're definitely right. I think it should be (if `dropna=False`):


| b       | 3           | 4  | All|
| ------------- |:-------------:| -----:|-----:|
|a|||
|1.0|1|0|1|
|2.0|1|3|4|
|np.nan|0|1|1|
|All|2|4|6|

- I need to fix the np.nan as it is still being ignored even with the current fix when `dropna=False` (i.e. the fix only works for `None`)
- Actually I didn't get why removing `dropna` here would help yet. I'll check closer.

**Comment by OXPHOS** [comment]
I think the change in `Cython` is definitely required. The problem is how to pass `dropna` to it without disturbing too many existing structures. 
I just reset my developing environment and am trying to use Anaconda Python2.7. Interestingly, numerous tests failed on me even with the master branch. So I just tested `pivot` and `groupby` at local. I'll do more research on the weekend.

**Comment by OXPHOS** [comment]
Some tests will be falling and many are actually different/separate problems. I already located several and will update soon.

**Comment by jreback** [comment]
can you rebase and update?

**Comment by jreback** [comment]
needs a rebase. if you'd like to continue, pls comment.

**Pull request closed without merging**